src/backend/joanie/tests/payment/test_models_credit_card.py

```diff
@@ -184,3 +186,109 @@ def test_models_credit_card_demote_a_credit_card_from_main_is_forbidden(self):
             str(context.exception),
             "{'__all__': ['Demote a main credit card is forbidden']}",
         )
+
+    def test_models_credit_card_delete_unused_1(self):
+        """
+        The manager method `delete_unused` should
+        unlink cards that are linked to orders that wont need payment anymore,
+        and delete all the credit cards that are not linked to any order.
+        """
+        for state, _ in enums.ORDER_STATE_CHOICES:
+            OrderFactory(state=state, product__price=10)
+
+        # With this order factory, we create a credit card for all states
+        self.assertEqual(CreditCard.objects.count(), len(enums.ORDER_STATE_CHOICES))
+
+        unlinked_credit_cards, deleted_credit_cards = CreditCard.objects.delete_unused()
+
+        # Order states that should not have a credit card linked
+        no_card_order_states = enums.ORDER_INACTIVE_STATES + (
+            enums.ORDER_STATE_COMPLETED,
+        )
+        self.assertEqual(len(unlinked_credit_cards), len(no_card_order_states))
+        self.assertEqual(len(deleted_credit_cards), len(no_card_order_states))
+
+        self.assertFalse(
+            Order.objects.filter(
+                state__in=no_card_order_states, credit_card__isnull=False
+            ).exists()
+        )
+        self.assertFalse(
+            CreditCard.objects.filter(orders__state__in=no_card_order_states).exists()
+        )
```

Comment:
I prefer this one as we add new state later this test will also take in account this one while with the second one we will have to think to update this test.